contracts/mainnet/connectors/compound-iii/helpers.sol

```diff
@@ -0,0 +1,169 @@
+		address to,
+		uint256 amt
+	) internal {
+		if (from == address(0)) {
+			CometInterface(market).transferAsset(to, token, amt);
+		} else {
+			CometInterface(market).transferAssetFrom(from, to, token, amt);
+		}
+	}
+
+	function _borrowOrWithdraw(
+		BorrowWithdrawParams memory params,
+		bool isWithdraw
+	) internal returns (uint256 amt, uint256 setId) {
+		uint256 amt_ = getUint(params.getId, params.amt);
+
+		require(
+			params.market != address(0) && params.token != address(0),
+			"invalid market/token address"
+		);
+		bool isEth = params.token == ethAddr;
+		address token_ = isEth ? wethAddr : params.token;
+
+		TokenInterface tokenContract = TokenInterface(token_);
+
+		uint256 initialBal = getAccountSupplyBalanceOfAsset(
+			address(this),
+			params.market,
+			token_
+		);
+
+		amt_ = amt_ == uint256(-1) ? initialBal : amt_;
+
+		if (isWithdraw) {
+			if (token_ == getBaseToken(params.market)) {
+				//from address is address(this) for withdrawOnBehalf
+				params.from = params.from == address(0)
+					? address(this)
+					: params.from;
+				uint256 balance = CometInterface(params.market).balanceOf(params.from);
+				if (balance > 0) {
+					require(
+						amt_ <= balance,
+						"withdraw-amt-greater-than-supplies"
+					);
+				}
+			}
+		} else {
+			params.from = params.from == address(0)
+				? address(this)
+				: params.from;
+			uint256 balance = CometInterface(params.market).balanceOf(params.from);
+
+			require(balance == 0, "borrow-disabled-when-supplied-base");
+		}
+
+		_withdraw(params.market, token_, params.from, params.to, amt_);
+
+		uint256 finalBal = getAccountSupplyBalanceOfAsset(
+			address(this),
+			params.market,
+			token_
+		);
+		amt_ = sub(initialBal, finalBal);
+
+		convertWethToEth(isEth, tokenContract, amt_);
+
+		setUint(params.setId, amt_);
+
+		amt = amt_;
+		setId = params.setId;
+	}
+
+	function getAccountSupplyBalanceOfAsset(
+		address account,
+		address market,
+		address asset
+	) internal returns (uint256 balance) {
+		if (asset == getBaseToken(market)) {
+			//balance in base
+			balance = CometInterface(market).balanceOf(account);
+		} else {
+			//balance in asset denomination
+			balance = uint256(
+				CometInterface(market).userCollateral(account, asset).balance
+			);
+		}
+	}
+
+	function setAmt(
+		address market,
+		address token,
+		address src,
+		uint256 amt,
+		bool isEth
+	) internal returns (uint256) {
+		if (isEth) {
+			if (amt == uint256(-1)) {
+				uint256 allowance_ = CometInterface(market).allowance(
+					src,
+					market
+				);
+				amt = src.balance < allowance_ ? src.balance : allowance_;
+			}
+			convertEthToWeth(isEth, TokenInterface(token), amt);
+		} else {
+			if (amt == uint256(-1)) {
+				uint256 allowance_ = CometInterface(market).allowance(
+					src,
+					market
+				);
+				uint256 bal_ = (token == getBaseToken(market))
+					? TokenInterface(market).balanceOf(src)
+					: CometInterface(market).userCollateral(src, token).balance;
```

Comment:
why this?

Reply:
This is also a general function, for the manager functions. 
Using it for transferAsset function so need to have collateral balance, and if the asset is base for the deposit case then the balance of base and adding one more condition as the borrow balance if it is for repayment.
Then the minimum of this balance and allowance is taken.

Reply:
Okay got it